r/R/dplyr-funcs-type.R

```diff
@@ -76,6 +76,60 @@ register_bindings_type_cast <- function() {
   register_binding("as.numeric", function(x) {
     Expression$create("cast", x, options = cast_options(to_type = float64()))
   })
+  register_binding("as.Date", function(x,
+                                       format = NULL,
+                                       tryFormats = "%Y-%m-%d",
+                                       origin = "1970-01-01",
+                                       tz = "UTC") {
+
+    if (call_binding("is.Date", x)) {
+      # base::as.Date() first converts to the desired timezone and then extracts
+      # the date, which is why we need to go through timestamp() first
+      return(x)
+
+    # cast from POSIXct
+    } else if (call_binding("is.POSIXct", x)) {
+      if (tz == "UTC") {
+        x <- build_expr("cast", x, options = cast_options(to_type = timestamp(timezone = tz)))
+      } else {
+        abort("`as.Date()` with a timezone different to 'UTC' is not supported in Arrow")
+      }
+
+    # cast from character
+    } else if (call_binding("is.character", x)) {
+      # this could be improved with tryFormats once strptime returns NA and we
+      # can use coalesce - https://issues.apache.org/jira/browse/ARROW-15659
+      # TODO revisit once https://issues.apache.org/jira/browse/ARROW-15659 is done
+      if (is.null(format)) {
+        if (length(tryFormats) == 1) {
+          format <- tryFormats[1]
+        } else {
+          abort("`as.Date()` with multiple `tryFormats` is not supported in Arrow yet")
+        }
+      }
+      x <- build_expr("strptime", x, options = list(format = format, unit = 0L))
+
+    # cast from numeric
+    } else if (call_binding("is.numeric", x)) {
+      # the origin argument will be better supported once we implement temporal
+      # arithmetic (https://issues.apache.org/jira/browse/ARROW-14947)
+      # TODO revisit once the above has been sorted
+      if (!call_binding("is.integer", x)) {
+        # Arrow does not support direct casting from double to date so we have
+        # to convert to integers first - casting to int32() would error so we
+        # need to use `floor` before casting. `floor` is also a bit safer than
+        # int32() with `safe = FALSE` since it doesn't switch to `ceiling` for
+        # negative numbers
+        # TODO revisit if arrow decides to support double -> date casting
+        x <- call_binding("floor", x)
+        x <- build_expr("cast", x, options = cast_options(to_type = int32()))
+      }
+      if (origin != "1970-01-01") {
+        abort("`as.Date()` with an `origin` different than '1970-01-01' is not supported in Arrow")
+      }
+    }
```

Comment:
Done. Moved all args checks at the top.

Reply:
Like I mentioned above, I honestly hadn't caught that this was only happening with numerics. We could put this at the top of the numeric if clause, that would be fine (or we could have a two-condition if at the top to check it). Though if someone _does_ provide an origin for anything other than numerics we probably should error since that's not going to do anything for them.

Reply:
I think you're right. However, base R doesn't complain if you provide arguments that are not necessary. Most are passed around via `...`.
``` r
as.Date("1981-01-01", origin = "1970-01-02")
#> [1] "1981-01-01"
```

<sup>Created on 2022-03-03 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Reply:
_nods_ yeah, so maybe we should keep it only when we're dealing with numerics. I'm doing another pass through the code this morning, so I'll comment on it in place. Thanks for this follow up!